### ISSUE TITLE
Deep neural networks for youtube recommendations

### ISSUE THREAD

**Issue by sooooner** [sooooner/papers#37]
# Deep neural networks for youtube recommendations[[link](https://static.googleusercontent.com/media/research.google.com/ko//pubs/archive/45530.pdf)]

### ABSTRACT
The paper is split according to the classic two-stage information retrieval dichotomy
1. 후보 생성 모델
2. 랭킹 모델

## INTRODUCTION
주요 도전 과제
1. scale : 방대한 양의 데이터 
2. Freshness : 새로운 영상에 대한 추천
3. Noise : 관측 불가능한 외부의 요소로 인해 좋은 ontology 필요

## SYSTEM OVERVIEW
두개의 NN으로 이루어져있다.

![image](https://user-images.githubusercontent.com/31437443/112978548-72ab1700-9192-11eb-8bc0-263552139013.png)

1. condidate generation

input : 유튜브 활동 기록
output : handreds condidate 
높은 정밀도로 사용자와 연관이 있음 CGNN(condidate generation neural network)은 CF(collaborative filtering)로 인해 개인화가 공급됨
CF에서 유사도는 : 영상 ID, 검색어, demographics와 같은 feature로 표현됨

2. ranking model

영상, 사용자의 특성 묘사의 집합을 사용한 목적함수를 통해 후보영상에 점수를 할당함
이 점수를 바탕으로 사용자에게 영상이 제공함

위 two-stage는 좋은 추천을 해주며 자른 정보와의 결합성도 좋음

개발 과정에서 여러 metric(recall, precision, ranking loss)를 사용해 범위를 줄이고 최종 단계에서 A/B live experiment사용
-> CRT, 시청시간, 사용자 참여등 추천 상목의 미묘한 변화를 추정가능(특히 A/B 는 항상 offline실험과 관계있는것만은 아니라서 중요)

## CANDIDATE GENERATION
이전에는 ranking loss를 사용한 MF기법을 사용했다.
MF를 적용한 NN방법은 초기에(오직 사용자의 시청기록을 임베딩) 얕은 NN MF 기법을 흉내내었다
-> 이런 관점에서 제안하는 방식은 Factorization 기법의 비선형 일반화 기법으로 볼수 있다.

### Recommendation as Classification
추천 시스템을 극단적인 분류문제로 바라보았다.
![image](https://user-images.githubusercontent.com/31437443/112978632-8c4c5e80-9192-11eb-871c-d00a4158ca97.png)
U : 사용자 "embedding"
C : context
V : 후보 영상 corpus "embedding"
Wt : 수백만의 비디오중 t 시간에 Vt를 시청하는 정확한 분류

task : softmax 분류기를 사용하여 영상을 구별하는데 유용한 (사용자의 기록, 맥락의 함수로써) 사용자 임배딩 학습

explicit feedback이 있지만 implicit feedback을 사용함
-> explicit이 매우 드믄 경우 implicit이 더 많은 정보와 좋은 추천을 제공함

#### Efficient Extreme Multiclass
효율적으로 training 하기 위해 Negative sampling을 사용하고 importance weighting으로 샘플링을 수정함
-> 크로스 앤트로피를 최소화 시키고 실제로 softmax보다 100배 이상읫 ㅗㄱ도를 보임(hierarchial softmax는 별로드라)

### Model Architecture
input : 영상의 embedding
사용자의 시청기록은 임배딩을 사용한 V표현에 mapping된 희소영상 ID의 가변길이 seq로 표현됨
NN은 고정길이의 dense V가 필요하기 떄문에 단순히 평균내서 사용함(여러 연산중에 성능이 젤 좋았다고 함)
embedding도 backprop을 통해 훈련됨

### Heterogeneous Signals
MF를 NN로 일반화 하는것의 핵심 장점은 연속, 범주인 feature를 모델에 쉽게 적용이 가능한다

검색 기록도 시청기록과 유사하게 처림(각 쿼리는 uni, bi-gram으로 토큰화 되어 임베딩됨)
embedding된 쿼리는 dense searching history로 요약됨

geographic feature는 새로운 사용자의 prior에 중요하게 작용함!
지역, 기기등 통계 정보도 embedding되고 합쳐쳐서 input으로 들어감

성별, 나이, 로그인 등등은 normalized해서 [0, 1] 제공됨

#### “Example Age” Feature(영상의 나이)
새로운 영상 추천은 매우 중요함, 사용자는 신선한 영상을 선호함
하지만ML을 그대로 적용하면 오래된 영상을 더 많이 추천함
따라서 훈련시에 영상의 나이를 넣음

![image](https://user-images.githubusercontent.com/31437443/112980285-a6873c00-9194-11eb-8b92-c424e55addde.png)

### Label and Context Selection
추천은 전형적인 surrogate problem(다른 문제로 추천 문제를 해결가능)
ex) 등급을 예측하면 좋은 영화 추천으로 이어짐

surrogate problem learning을 선택하는 것이 매우 중요함(A/B test에서) 하지만 offline 실험으로 추정이 어려움

유튜브 시청 이력 말고도 다른 사이트의 시청 목록도 활용함
유튜브 내의 시청 기록은 이미 추천 영상 이기 때문에 치우친 데이터 이기도함

사용자가 추천 이외에 방법으로 영상을 발견했다면 CF로 이 **'발견'**을 신속하게 전파하기를 원함

live metric 개선의 또다른 방안 : 사용자랑 고정된 수의 예제를 생성해 loss function에서 사용자에게 효과적으로 가중치를 부여하는것
-> 소규모 유튜브 폐인의 loss 지배를 방지 할수 있음

다소 직관적이제 **withhold information from the classifier**는 중요하다
ex) '테일러 스위프트'를 검색했다면 바로 다음 시청할 비디오로 계속 '테일러 스위프트'가 나온다
-> bag-of-token 사용

사용자의 영상 감삼 패턴은 비대칭이다
ex) 에피소드 1-2-3-4, 음악: 유명-> 마이너
CF에서는 유저 기록과 랜덤 아이템을 유지한다
-> 소비 패턴 무시 -> 효과 안좋음

###  Experiments with Features and Depth 
Features and Depth를 추가 하면 성능이 상승함
실험에서 영상 100만, 검색어 100만을 256차원 embedding, bag-size : 최근 50비디어(검색)

![image](https://user-images.githubusercontent.com/31437443/112982177-22828380-9197-11eb-95bb-7ede77255dc4.png)

## RANKING
ranking에서 사용자가 받는 **"인상"**은 매우 중요함
사용자는 일반적으로 주어지는 영상을 보겠지만! 썸네일 어그로에 걸릴수도 있음

랭킹을 매길때 후보 모두 계산하는것이 아니기 때문에 영상/사용자와 영상의 관게를 설명하는 더 많은 feature access가 있다!

\* 랭킹은 점수가 직접적으로 비교할 수 없는 서로다른 후보 소스 조합이 중요

RANKING도 후보 생성 구조와 비슷한 구조를 사용할거 + logistic regression을 사용해서 video impression score 할당
그런다음 유저에게 제공됨
A/B test로 지속적인 조정을 하지만 일번적으로 노출당 예상 시청 시간의 간단한 함수임

### Feature Representation
univalent : impression video id
multivalent : history video id

#### Feature Engineering
Feature Engineering에 많은 노력을 함
주요 challenge
1. 사용자의 이용 패턴
2. 여러 정보가 어떻게 관계 맺고 있는지

\* 중요한 신호는 다른 사용자의 ranking experience와 사용자의 item이나 다른 item관의 이런 상호 작용의 매칭 @.@
ex) 이 채널에서 사용자가 몇개의 영상을 봤는지, 마지막으로 이 주제릐 영상을 본게 언제인지...
-> 이는 항목간의 일반화가 잘 되어서 특히 강력하다

\* 후보 생성에서 정보를 feature 형태로 ranking에 전파하는것도 중요
1. 왜 이영상이 추천 되었는지 또 어떤 점수로 할당 했는지
2. "인상적인 비디오"의 빈도(추천했는데 안본건 강등)

#### Embedding Categorical Features
log(ID space dimension) 비례 embedding dimension

embedding된 vocab은 한번 훈련된수 look-up table이 됨
Top-N 이 외에는 0 mapping

동일한 ID 공간의 범주는 기본 임배딩을 공유함
영상 ID, 마지막 영상 ID, seed ID 전부 사용
- 별도의 NN에 들어감(다른 역할 이니깐)
- 
#### Normalizing Continuous Features
x -> [0, 1] scaling

![image](https://user-images.githubusercontent.com/31437443/112983034-24991200-9198-11eb-8c18-1232e43b2cd6.png)

## CONCLUSIONS
YouTube 추천을 위한 두가지 모델 설명, feature engineering, modeling
